layouts/company/single.html

```diff
@@ -102,20 +102,33 @@
         {{ T "cdb-help-needed" (dict "editUrl" (print ("suggest" | absURL) "?type=edit&for=cdb&slug=" .Params.slug) "newUrl" (print ("suggest" | absURL) "?type=new&for=cdb")) | safeHTML }}
     </div>
 
-    <!-- TODO: This is currently a very naive approach, mostly since we don't exactly have any data to do this better. We will want to improve this in the future. Oh yeah, and this feature seems to be very 'non-deterministic', at least under our parameters. Not sure if that is a good or a bad thing. -->
+    <!-- TODO: This is currently a very naive approach, mostly since we don't exactly have any data to do this better. We will want to improve this in the future. -->
     {{ if not .Site.Params.devMode }}
-    {{ $related := (where (.Site.RegularPages.Related .) "Type" "company") | first 5 }}
-    {{ with $related }}
-    <div id="related-companies">
-        <h2>{{ T "cdb-related-companies" }}</h2>
-        <!-- TODO: Nicer styling -->
-        <ul>
-            {{ range . }}
-            <li><a href="{{ .Permalink }}">{{ .Param "name" }}</a></li>
+        {{ $related_pages := slice }}
+
+        {{ $num_categories := (cond (isset .Params "categories") .Params.categories slice) | len}}
+        {{ $num_countries := (.Param "relevant-countries") | len}}
+
+        {{ range .Site.Pages }}
+            {{ $num_common_categories := intersect $.Params.categories .Params.categories | len }}
+            {{ $num_common_countries := intersect ($.Param "relevant-countries") (.Param "relevant-countries") | len }}
+            {{ if and (gt $num_common_categories 0) (gt $num_common_countries 0) (ne .Permalink $.Permalink) }}
+                {{ $score := div (add (div (float $num_common_categories) $num_categories) (div $num_common_countries $num_countries)) 2 }}
+                {{ $related_pages = $related_pages | append (dict "page" . "score" $score) }}
             {{ end }}
-        </ul>
-    </div>
-    {{ end }}
+        {{ end }}
+        {{ $related_pages = sort $related_pages "score" "desc" }}
```

Comment:
Also, couldn't this be done in the first loop to save time?

Reply:
Not really sure how this would work to be honest. Running the build with and without this line only causes time differences well within the margin of error though. We are talking ±200 ms, with the winner alternating between the two, so it doesn't seem to make any significant difference.

Reply:
Interestingly enough, changing the if clause to `{{ if and (gt $num_common_categories 0) (ne .Permalink $.Permalink) }}` consistently decreases build times by about a second, though.

That would mean showing companies with no related countries if there are none with related countries. I actually think I like that better anyway. What do you think, @zner0L?

Reply:
There is also no actual reason to normalize the scores to [0, 1]. Removing that doesn't cause a measurable change in build times, though. Your call.

Reply:
I think we should do that then. Normalization can stay in there.